From 02d0a17e223735d262d0188d4c84c8a53a2b7b8a Mon Sep 17 00:00:00 2001
From: Dylan Yip <dylan.yip@xilinx.com>
Date: Mon, 26 Aug 2019 15:25:59 -0700
Subject: [PATCH 0662/1852] v4l: xilinx: scd: Proper unregister from V4L2
 framework

commit 7c34a8eb886ede38f7d3a7be80181cdc4b7ca5dc from
https://github.com/Xilinx/linux-xlnx.git

This patch fixes cleanup of scd device. It properly unregisters
subdev channels.

Signed-off-by: Dylan Yip <dylan.yip@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../xilinx/xilinx-scenechange-channel.c        | 18 ++++++++++++++++++
 .../media/platform/xilinx/xilinx-scenechange.c |  7 +++++++
 .../media/platform/xilinx/xilinx-scenechange.h |  2 ++
 3 files changed, 27 insertions(+)

diff --git a/drivers/media/platform/xilinx/xilinx-scenechange-channel.c b/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
index 39833f753274..b0270ba848b4 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
+++ b/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
@@ -420,3 +420,21 @@ int xscd_chan_init(struct xscd_device *xscd, unsigned int chan_id,
 	mutex_destroy(&chan->lock);
 	return ret;
 }
+
+/**
+ * xscd_chan_cleanup - Clean up the V4L2 subdev for a channel
+ * @xscd: Pointer to the SCD device structure
+ * @chan_id: Channel id
+ * @node: device node
+ */
+void xscd_chan_cleanup(struct xscd_device *xscd, unsigned int chan_id,
+		       struct device_node *node)
+{
+	struct xscd_chan *chan = &xscd->chans[chan_id];
+	struct v4l2_subdev *subdev = &chan->subdev;
+
+	v4l2_async_unregister_subdev(subdev);
+	v4l2_ctrl_handler_free(&chan->ctrl_handler);
+	media_entity_cleanup(&subdev->entity);
+	mutex_destroy(&chan->lock);
+}
diff --git a/drivers/media/platform/xilinx/xilinx-scenechange.c b/drivers/media/platform/xilinx/xilinx-scenechange.c
index 9135355934fe..5d92b24b3eae 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange.c
+++ b/drivers/media/platform/xilinx/xilinx-scenechange.c
@@ -162,6 +162,13 @@ static int xscd_probe(struct platform_device *pdev)
 static int xscd_remove(struct platform_device *pdev)
 {
 	struct xscd_device *xscd = platform_get_drvdata(pdev);
+	struct device_node *subdev_node;
+	unsigned int id = 0;
+
+	for_each_child_of_node(xscd->dev->of_node, subdev_node) {
+		xscd_chan_cleanup(xscd, id, subdev_node);
+		id++;
+	}
 
 	xscd_dma_cleanup(xscd);
 	clk_disable_unprepare(xscd->clk);
diff --git a/drivers/media/platform/xilinx/xilinx-scenechange.h b/drivers/media/platform/xilinx/xilinx-scenechange.h
index 9749a7784bbb..d5dfaa02b5e0 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange.h
+++ b/drivers/media/platform/xilinx/xilinx-scenechange.h
@@ -237,4 +237,6 @@ void xscd_dma_cleanup(struct xscd_device *xscd);
 void xscd_chan_event_notify(struct xscd_chan *chan);
 int xscd_chan_init(struct xscd_device *xscd, unsigned int chan_id,
 		   struct device_node *node);
+void xscd_chan_cleanup(struct xscd_device *xscd, unsigned int chan_id,
+		       struct device_node *node);
 #endif
-- 
2.31.1

